src/notification.rs

```diff
@@ -1,218 +1,163 @@
-use once_cell::sync::Lazy;
-use url::Url;
+use std::collections::BTreeMap;
+use std::fmt::Write;
 
-static DEFAULT_MUSTACHE_TEMPLATE: Lazy<mustache::Template> = Lazy::new(|| {
-    mustache::compile_str(
-        "
-{{#singlehost}}
-{{.}} changed
-{{/singlehost}}
-{{^singlehost}}
-{{siteamount}} websites changed
-{{/singlehost}}
-
-{{#sites}}
-- {{.}}
-{{/sites}}
-
-{{#commit}}
-See {{.}}
-{{/commit}}
-",
-    )
-    .unwrap()
-});
-
-#[deprecated = "The notification feature will be removed"]
-#[derive(serde::Serialize)]
-pub struct MustacheData {
-    commit: Option<String>,
-    #[deprecated = "use singlehost"]
-    singledomain: Option<String>,
-    singlehost: Option<String>,
-    siteamount: usize,
-
-    #[deprecated = "use hosts"]
-    domains: Vec<String>,
-    hosts: Vec<String>,
-    sites: Vec<Url>,
-}
+use url::Url;
 
-impl MustacheData {
-    pub fn new(commit: Option<String>, changed_urls: Vec<Url>) -> Self {
-        let mut sites = changed_urls;
-        sites.sort_unstable();
-        sites.dedup();
+fn generate_change_lines(mut changed: Vec<Url>) -> String {
+    changed.sort_unstable();
+    changed.dedup();
 
-        let mut hosts = sites
-            .iter()
-            .filter_map(Url::host_str)
-            .map(std::string::ToString::to_string)
-            .collect::<Vec<_>>();
-        hosts.dedup();
-
-        let singlehost = if let [single] = hosts.as_slice() {
-            Some(single.clone())
-        } else {
-            None
-        };
-
-        Self {
-            commit,
-            singledomain: singlehost.clone(),
-            singlehost,
-            siteamount: sites.len(),
-
-            domains: hosts.clone(),
-            hosts,
-            sites,
+    let mut changed_hosts = BTreeMap::<String, Vec<Url>>::new();
+    for url in changed {
+        if let Some(host) = url.host_str() {
+            let host = host.to_owned();
+            changed_hosts.entry(host).or_default().push(url);
         }
     }
 
-    pub fn apply_to_template(
-        &self,
-        template: Option<&mustache::Template>,
-    ) -> anyhow::Result<String> {
-        let template = template.unwrap_or_else(|| &DEFAULT_MUSTACHE_TEMPLATE);
-        Ok(template.render_to_string(self)?.trim().to_owned())
-    }
-
-    fn example_single(commit: Option<&str>) -> Self {
-        Self::new(
-            commit.map(ToOwned::to_owned),
-            vec![Url::parse("https://edjopato.de/post/").unwrap()],
-        )
-    }
+    let mut text = String::new();
 
-    fn example_different(commit: Option<&str>) -> Self {
-        Self::new(
-            commit.map(ToOwned::to_owned),
-            vec![
-                Url::parse("https://edjopato.de/post/").unwrap(),
-                Url::parse("https://foo.bar/").unwrap(),
-            ],
-        )
+    for urls in changed_hosts.values() {
+        if let [single_url] = &**urls {
```

Comment:
What does this mean? Are you trying to filter the host that only have one url?

Reply:
Exactly. I regularly use slice matching like this. Especially with the `urls.len() > 1` in the next for loop it seems relatively clear to me.

`&*urls` is the equivalent of `&urls[..]` with a deep reference here (`values()` returns a reference to a reference so it needs multiple `*` to get out of it (Which to me personally is the only annoying part of this line). The long form if this would be [`Vec::as_slice`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_slice).

`if let [one] = slice` is only true when the slice has exactly one element. I regularly use such slice checks (also in match statements) and they seem readable to me even without an explicit `as_slice` (as it has to be slice then anyway).

Reply:
I hope to achieve easier to read lists with that. Currently all URL are listed without any host grouping which would be the simple option.